### PR TITLE
[8/n] Migrate merge_attn_states, mamba, sampler to torch stable ABI (continued)

### PR DESCRIPTION
**Pull request by cleonard530**
Continuation of PR #38841

@janeyx99 

Stacked on #43209, commits to review https://github.com/vllm-project/vllm/pull/43361/changes/2292383dce4e96a468df7d8aac22315c114641a6..dab4bae2e67327dfe418ada22a20f54cdf871e6b

-------------------------------------------------------------------------------------------
Commits to review

## Test Plan
```
pytest tests/kernels/attention/test_merge_attn_states.py
pytest tests/kernels/test_top_k_per_row.py
pytest tests/kernels/test_apply_repetition_penalties.py
pytest tests/kernels/mamba/test_mamba_ssm.py
```

## Test Result
Screenshot 2026-04-02 at 4 23 36 PM
Screenshot 2026-04-02 at 4 23 57 PM

Screenshot 2026-04-03 at 1 43 21 PM

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

------------------------------------------------------------------------------------

**Differences between this and PR #38841**
- The merge conflicts that were fixed. These will be pointed out in the comments below. 
- There was a big divergence between that PR and main for the commit to migrate sampler kernels and topk, so I did that commit myself instead of using Mikaya's from the previous PR. These differences mostly show up in the `csrc/libtorch_stable/topk.cu` file. 
- Because topk now used persistent_topk, I needed to also update `csrc/persistent_topk.cuh`
- Had to add sharedMemPerBlockOptin = "sharedMemPerBlockOptin" to pyproject.toml to bypass pre-commit typo error. 

**Migration progress using the Audit Python extension [torch-abi-audit](https://pypi.org/project/torch-abi-audit/):**

  **main branch:**
```
  -- extensions --
      [UNSTABLE] [abi3-ok               ] _C.abi3.so  (stable_shim=0, unstable=99)
      [STABLE  ] [abi3-ok               ] _C_stable_libtorch.abi3.so  (stable_shim=78, unstable=0)
      [UNSTABLE] [abi3-ok               ] _flashmla_C.abi3.so  (stable_shim=0, unstable=72)
      [UNSTABLE] [abi3-ok               ] _flashmla_extension_C.abi3.so  (stable_shim=0, unstable=68)
      [UNSTABLE] [abi3-ok               ] _moe_C.abi3.so  (stable_shim=0, unstable=81)
      [NO-TORCH] [abi3-ok               ] cumem_allocator.abi3.so
      [NO-TORCH] [abi3-ok               ] spinloop.abi3.so
      [UNSTABLE] [uses-private-api      ] third_party/deep_gemm/_C.cpython-312-x86_64-linux-gnu.so  (stable_shim=0, unstable=59)
      [UNSTABLE] [uses-private-api      ] third_party/deep_gemm/_C.cpython-313-x86_64-linux-gnu.so  (stable_shim=0, unstable=59)
      [UNSTABLE] [abi3-ok               ] vllm_flash_attn/_vllm_fa2_C.abi3.so  (stable_shim=0, unstable=85)
      [UNSTABLE] [abi3-ok               ] vllm_flash_attn/_vllm_fa3_C.abi3.so  (stable_shim=0, unstable=81)

```

  **this branch:**
```
   -- extensions --
    [UNSTABLE] [abi3-ok               ] _C.abi3.so  (stable_shim=0, unstable=98)
    [STABLE  ] [abi3-ok               ] _C_stable_libtorch.abi3.so  (stable_shim=78, unstable=0)
    [UNSTABLE] [abi3-ok               ] _flashmla_C.abi3.so  (stable_shim=0, unstable=72)
    [UNSTABLE] [abi3-ok               ] _flashmla_extension_C.abi3.so  (stable_shim=0, unstable=68)
    [UNSTABLE] [abi3-ok               ] _moe_C.abi3.so  (stable_shim=0, unstable=81)
    [NO-TORCH] [abi3-ok               ] cumem_allocator.abi3.so
    [NO-TORCH] [abi3-ok               ] spinloop.abi3.so
    [UNSTABLE] [uses-private-api      ] third_party/deep_gemm/_C.cpython-312-x86_64-linux-gnu.so  (stable_shim=0, unstable=59)
    [UNSTABLE] [uses-private-api      ] third_party/deep_gemm/_C.cpython-313-x86_64-linux-gnu.so  (stable_shim=0, unstable=59)
    [UNSTABLE] [abi3-ok               ] vllm_flash_attn/_vllm_fa2_C.abi3.so  (stable_shim=0, unstable=85)
    [UNSTABLE] [abi3-ok               ] vllm_flash_attn/_vllm_fa3_C.abi3.so  (stable_shim=0, unstable=81)

```
      
 i.e. 1 unstable extension was removed from _C.abi.so.

### CODE REVIEW

File: csrc/libtorch_stable/ops.h [hunk 1]
```diff
@@ -220,6 +231,48 @@ void fused_qk_norm_rope(torch::stable::Tensor& qkv, int64_t num_heads_q,
...
+void top_k_per_row_decode(const torch::stable::Tensor& logits, int64_t next_n,
+                          const torch::stable::Tensor& seqLens,
+                          torch::stable::Tensor& indices, int64_t numRows,
+                          int64_t stride0, int64_t stride1, int64_t topK);
+
+void persistent_topk(const torch::stable::Tensor& logits,
+                     const torch::stable::Tensor& lengths,
+                     torch::stable::Tensor& output,
+                     torch::stable::Tensor& workspace, int64_t k,
+                     int64_t max_seq_len);
```
Comment (cleonard530) [inline csrc/libtorch_stable/ops.h:255]: This changed from PR #38841, where it used to be large_context_topk. There were also a lot of other kernels added in this file in that PR that were not migrated over so they were left out.

File: csrc/libtorch_stable/sampler.cu [hunk 2]
```diff
@@ -645,27 +643,31 @@ void apply_repetition_penalties_(
...
+void top_k_per_row_decode(const torch::stable::Tensor& logits, int64_t next_n,
+                          const torch::stable::Tensor& seqLens,
+                          torch::stable::Tensor& indices, int64_t numRows,
+                          int64_t stride0, int64_t stride1, int64_t topK) {
   constexpr int kSortingAlgorithmThreshold = 12288;
   constexpr int kSplitWorkThreshold = 200 * 1000;
   constexpr int kNumThreadsPerBlock = 512;
-  const cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  const torch::stable::accelerator::DeviceGuard device_guard(
+      logits.get_device_index());
```
Comment (cleonard530) [inline csrc/libtorch_stable/sampler.cu]: This guard was added in this PR but not in the other

File: csrc/libtorch_stable/sampler.cu [hunk 3]
```diff
@@ -677,73 +679,78 @@ void top_k_per_row_decode(const torch::Tensor& logits, int64_t next_n,
...
 
     vllm::topKPerRowDecode<kNumThreadsPerBlock, true, true>
         <<<dim3(numRows, multipleBlocksPerRowConfig), kNumThreadsPerBlock,
            2 * topK * sizeof(int32_t), stream>>>(
-            logits.data_ptr<float>(), seqLens.data_ptr<int>(),
-            outIndicesAux.data_ptr<int>(), static_cast<int>(stride0),
+            logits.const_data_ptr<float>(), seqLens.const_data_ptr<int>(),
+            outIndicesAux.mutable_data_ptr<int>(), static_cast<int>(stride0),
             static_cast<int>(stride1), static_cast<int>(topK),
             static_cast<int>(next_n), seqLensIs2D,
```
Comment (cleonard530) [inline csrc/libtorch_stable/sampler.cu:709]: Slightly different than other PR b/c seqLensIs2D was added

File: csrc/libtorch_stable/sampler.cu [hunk 4]
```diff
@@ -677,73 +679,78 @@ void top_k_per_row_decode(const torch::Tensor& logits, int64_t next_n,
...
     vllm::topKPerRowDecode<kNumThreadsPerBlockMerge, true, false, true>
         <<<numRows, kNumThreadsPerBlockMerge, topK * sizeof(int32_t), stream>>>(
-            outLogitsAux.data_ptr<float>(), seqLens.data_ptr<int>(),
-            indices.data_ptr<int>(), multipleBlocksPerRowConfig * topK, 1,
-            static_cast<int>(topK), static_cast<int>(next_n), seqLensIs2D,
-            nullptr, multipleBlocksPerRowConfig, outIndicesAux.data_ptr<int>());
+            outLogitsAux.const_data_ptr<float>(), seqLens.const_data_ptr<int>(),
+            indices.mutable_data_ptr<int>(), multipleBlocksPerRowConfig * topK,
+            1, static_cast<int>(topK), static_cast<int>(next_n), seqLensIs2D,
+            nullptr, multipleBlocksPerRowConfig,
```
Comment (cleonard530) [inline csrc/libtorch_stable/sampler.cu:718]: Same as above

File: csrc/libtorch_stable/sampler.cu [hunk 5]
```diff
@@ -677,73 +679,78 @@ void top_k_per_row_decode(const torch::Tensor& logits, int64_t next_n,
...
+void top_k_per_row_prefill(const torch::stable::Tensor& logits,
+                           const torch::stable::Tensor& rowStarts,
+                           const torch::stable::Tensor& rowEnds,
+                           torch::stable::Tensor& indices, int64_t numRows,
+                           int64_t stride0, int64_t stride1, int64_t topK) {
   constexpr int kSortingAlgorithmThreshold = 12288;
   constexpr int kNumThreadsPerBlock = 512;
-  const cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  const torch::stable::accelerator::DeviceGuard device_guard(
+      logits.get_device_index());
```
Comment (cleonard530) [inline csrc/libtorch_stable/sampler.cu]: Added guard

File: csrc/libtorch_stable/topk.cu [hunk 6]
```diff
@@ -1,49 +1,51 @@
 // Persistent TopK kernel for DeepSeek V3 sparse attention indexer.
```
Comment (cleonard530) [inline csrc/libtorch_stable/topk.cu:1]: This file s going to look very different from #38841 because of big changes on main.

File: csrc/ops.h [hunk 7]
```diff
@@ -75,26 +68,6 @@ void fused_deepseek_v4_qnorm_rope_kv_rope_quant_insert(
...
-
-void top_k_per_row_decode(const torch::Tensor& logits, int64_t next_n,
-                          const torch::Tensor& seqLens, torch::Tensor& indices,
-                          int64_t numRows, int64_t stride0, int64_t stride1,
-                          int64_t topK);
-
-void persistent_topk(const torch::Tensor& logits, const torch::Tensor& lengths,
-                     torch::Tensor& output, torch::Tensor& workspace, int64_t k,
-                     int64_t max_seq_len);
-
```
Comment (cleonard530) [inline csrc/ops.h:98]: Slightly different because the other PR used large_context_topk instead of persistent_topk

File: csrc/persistent_topk.cuh [hunk 8]
```diff
@@ -126,10 +126,10 @@ struct RadixRowState {
 // ============================================================================
```
Comment (cleonard530) [inline csrc/persistent_topk.cuh:127]: This file was not in the other PR but a few changes had to be made to it for `topk.cu` to work

File: pyproject.toml [hunk 9]
```diff
@@ -138,6 +138,7 @@ bbc5b7ede = "bbc5b7ede"
 NOOPs = "NOOPs"
 nin_shortcut = "nin_shortcut"
 cudaDevAttrMaxSharedMemoryPerBlockOptin = "cudaDevAttrMaxSharedMemoryPerBlockOptin"
+sharedMemPerBlockOptin = "sharedMemPerBlockOptin"
```
Comment (cleonard530) [inline pyproject.toml:144]: Had to add this here because the pre-commit linting check was giving a 'typo' warning for Optin

**Comment by Harry-Chen** [comment]
@cleonard530 Please fix the merge conflicts, I think it might come from your 7th PR?

File: csrc/libtorch_stable/topk.cu [hunk 10]
```diff
@@ -219,37 +223,45 @@ void launch_persistent_topk(const torch::Tensor& logits,
...
-              "persistent_topk supports k=512, k=1024, or k=2048, got k=", k);
+  STD_TORCH_CHECK(lengths.numel() == num_rows, "lengths size mismatch");
+  STD_TORCH_CHECK(output.size(0) == num_rows && output.size(1) == k,
+                  "output size mismatch");
+  STD_TORCH_CHECK(
+      k == 512 || k == 1024 || k == 2048,
+      "persistent_topk supports k=512, k=1024, or k=2048, got k=", k);
+
+  const torch::stable::accelerator::DeviceGuard device_guard(
+      logits.get_device_index());
```
Comment (cleonard530) [inline csrc/libtorch_stable/topk.cu]: New device guard

File: csrc/libtorch_stable/topk.cu [hunk 11]
```diff
@@ -131,27 +133,29 @@ void launch_persistent_topk(const torch::Tensor& logits,
...
           vllm::FilteredTopKRaggedTransform<float, int32_t, TopK>(
-              logits.data_ptr<float>(), output.data_ptr<int32_t>(),
-              lengths.data_ptr<int32_t>(), static_cast<uint32_t>(num_rows),
-              static_cast<uint32_t>(TopK), static_cast<uint32_t>(stride),
-              stream);
-      TORCH_CHECK(status == cudaSuccess,
-                  "FilteredTopK fallback failed: ", cudaGetErrorString(status));
+              logits.const_data_ptr<float>(),
+              output.mutable_data_ptr<int32_t>(),
+              lengths.const_data_ptr<int32_t>(),
```
Comment (cleonard530) [inline csrc/libtorch_stable/topk.cu:148]: lengths may need to be mutable, will take a closer look.

**Review by janeyx99 — APPROVED** [review]
topk looks good to me. please clarify the const-changes for lengths and input in the PR comment--I believe the const is correct (and matches the schema).

remove the deviceguards--i also think they're likely better but let's keep this change as modular as possible

File: csrc/libtorch_stable/ops.h (hunk 1 above)
Comment (janeyx99) [reply csrc/libtorch_stable/ops.h:255]: The other kernels are just from previous commits

File: csrc/persistent_topk.cuh [hunk 12]
```diff
@@ -126,10 +126,10 @@ struct RadixRowState {
...
 
 struct PersistentTopKParams {
-  const float* __restrict__ input;  // [num_rows, stride]
-  int32_t* __restrict__ output;     // [num_rows, top_k]
-  int32_t* __restrict__ lengths;    // [num_rows]
-  RadixRowState* row_states;        // large path: per-group state
+  const float* __restrict__ input;      // [num_rows, stride]
+  int32_t* __restrict__ output;         // [num_rows, top_k]
+  const int32_t* __restrict__ lengths;  // [num_rows]
```
Comment (cleonard530) [inline csrc/persistent_topk.cuh:131]: This variable `lengths` was changed to const since it is read-only.

File: csrc/persistent_topk.cuh [hunk 13]
```diff
@@ -1269,9 +1269,11 @@ constexpr int ComputeFilteredTopKVecSize(uint32_t max_len) {
 }
 
 template <typename DType, typename IdType, uint32_t MAX_K = 2048>
-cudaError_t FilteredTopKRaggedTransform(DType* input, IdType* output_indices,
-                                        IdType* lengths, uint32_t num_rows,
-                                        uint32_t top_k_val, uint32_t max_len,
+cudaError_t FilteredTopKRaggedTransform(const DType* input,
+                                        IdType* output_indices,
+                                        const IdType* lengths,
+                                        uint32_t num_rows, uint32_t top_k_val,
```
Comment (cleonard530) [inline csrc/persistent_topk.cuh:1275]: This variables `input` and  `lengths` were changed to const since they are read-only.

**Comment by cleonard530** [comment]
@Harry-Chen, Can you add the 'ready' label to this PR?

**Comment by cleonard530** [comment]
@Harry-Chen, I don't think any of these errors are due to the migration to the Torch stable ABI.

**Review by Harry-Chen — APPROVED** [review]
